### DISCUSSION TITLE
How to rebind / remap keys?

### DISCUSSION THREAD

**Discussion by alchem1no** [NeoVintageous/NeoVintageous#1024] [Q&A]
I want to rebind **esc** key to **tab** and **hjkl** keys to **jkl;** respectively. How can I do it in this plugin?

**Comment by gerardroche** [comment]
There is a lot of information about mapping here: https://neovintageous.github.io/reference/key-mapping

The [Colemak](https://github.com/gerardroche/NeoVintageousColemak/blob/master/colemak.neovintageous) and [Dvorak](https://github.com/gerardroche/NeoVintageousDvorak) packages have some example mappings.

Also my own: https://github.com/gerardroche/sublimefiles/blob/master/User/.neovintageousrc

Now, mapping Esc is a special case. Esc is treated specially in various contexts. I think the most robust way to map that key is to define a sublime key binding. There is how NeoVintageous defines the Esc and Tab key bindings:

```
    {"keys": ["escape"],    "command": "nv_feed_key", "args": {"key": "<esc>"},        "context": [{"key": "nv_command_or_insert"},  {"key": "nv_handle_key", "operand": "<Esc>"}]},
    {"keys": ["tab"],       "command": "nv_cmdline_feed_key", "args": {"key": "<tab>"},    "context": [{"key": "setting._nv_ex_mode"}]},
```

If you want to make Esc mean Tab and vice-versa you could try:

```
    {"keys": ["tab"],    "command": "nv_feed_key", "args": {"key": "<esc>"},        "context": [{"key": "nv_command_or_insert"},  {"key": "nv_handle_key", "operand": "<Esc>"}]},
    {"keys": ["escape"],       "command": "nv_cmdline_feed_key", "args": {"key": "<tab>"},    "context": [{"key": "setting._nv_ex_mode"}]},
``` 

This effectively "tricks" NeoVintageous into thinking it's receiving the alternate keys. Internally NeoVintageous will do the "feed key" action.

One thing to remember, is your `.neovintageousrc` mappings will receive the feed key, so if you are also mapping the keys in your rc file, you need to map the feed key.
